### PR TITLE
Another try at preponing the interpreter callbacks

### PR DESCRIPTION
**Pull request by Teemperor**

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/Interpreter.cpp [hunk 1]
```diff
@@ -210,6 +224,8 @@ namespace cling {
     // the lexer on EOF token.
     PP.enableIncrementalProcessing();
 
+    bool usingCxxModules = getSema().getLangOpts().Modules;
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/Interpreter.cpp]: Could we move that closer to its use?

File: interpreter/cling/lib/Interpreter/Interpreter.cpp [hunk 2]
```diff
@@ -226,6 +242,21 @@ namespace cling {
     DiagnosticConsumer& DClient = getCI()->getDiagnosticClient();
     DClient.BeginSourceFile(getCI()->getLangOpts(), &PP);
 
+    if (m_IncrParser->getCI()->getLangOpts().Modules) {
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/Interpreter.cpp]: Why not using `usingCxxModules` and moving the next if statement here?

File: interpreter/cling/lib/Interpreter/Interpreter.cpp [hunk 3]
```diff
@@ -237,6 +268,13 @@ namespace cling {
       return;
     }
 
+    // When not using C++ modules, we now have a PCH and we can safely setup
+    // our callbacks without fearing that they get ovewritten by clang code.
+    // The modules setup is handled above.
+    if (!usingCxxModules) {
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/Interpreter.cpp]: Can that be an else to the `if (cxxmodules)` case?

File: interpreter/cling/lib/Interpreter/Interpreter.cpp (hunk 1 above)
Comment (Teemperor) [reply interpreter/cling/lib/Interpreter/Interpreter.cpp]: Done.

File: interpreter/cling/lib/Interpreter/Interpreter.cpp (hunk 2 above)
Comment (Teemperor) [reply interpreter/cling/lib/Interpreter/Interpreter.cpp]: DOne.

File: interpreter/cling/lib/Interpreter/Interpreter.cpp (hunk 3 above)
Comment (Teemperor) [reply interpreter/cling/lib/Interpreter/Interpreter.cpp]: Well, the whole point is (as described in the comment) to have it before/after the `Initialize` call which overwrites the PCH in the non-module case.

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/Interpreter.cpp]: Ah, that's right. Let's see if the bots are green and I will merge it in.

**Pull request merged**